pkg/planner/core/plan_cost_ver2.go

```diff
@@ -184,7 +184,7 @@ func (p *PhysicalTableScan) GetPlanCostVer2(taskType property.TaskType, option *
 		}
 		hasFullRangeScan := ranger.HasFullRange(p.Ranges, unsignedIntHandle)
 
-		shouldApplyPenalty := hasFullRangeScan && preferRangeScanCondition
+		shouldApplyPenalty := hasFullRangeScan && (sessionVars.StmtCtx.GetHasForce() || preferRangeScanCondition)
```

Comment:
So our assumption is that if the user is using `use_index` hint, we assume they want `range-scan`?

Reply:
Better to add some comments if I understand this correctly.

Reply:
Done